File: haskell/src/Vectorize/IntrinsicBuilder.hs
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE OverloadedStrings #-}

module IntrinsicBuilder where

import qualified LLVM.AST as AST
import qualified LLVM.AST.Constant as C
import qualified LLVM.IRBuilder.Module as IRB
import qualified LLVM.IRBuilder.Instruction as IRB
import Data.Text.Format (format)
import Data.Text (Text)
import qualified Data.Map as Map
import Control.Monad.State

data IntrinsicBuilder = IntrinsicBuilder
  { instWrappers :: AST.Module
  , builder :: IRB.IRBuilder ()
  }

create :: IntrinsicBuilder -> Text -> [AST.Operand] -> Word8 -> IRB.IRBuilder AST.Operand
create IntrinsicBuilder{..} name operands imm8 = do
  let wrapperName = format "intrinsic_wrapper_{}_{}" (name, imm8)
  f <- case Map.lookup wrapperName (AST.moduleDefinitions instWrappers) of
         Just (AST.GlobalDefinition f@AST.Function{}) -> return f
         _ -> error "Intrinsic wrapper undefined."
  
  let [bb] = AST.basicBlocks f
  let numArgs = length $ AST.parameters f
  assert (length operands == numArgs) "Incorrect number of operands"

  vMap <- foldM (\m (arg, op) -> do
    assert (IRB.castIsValid IRB.BitCast op (AST.typeOf arg)) "Invalid input type"
    operand <- IRB.bitcast op (AST.typeOf arg)
    return $ Map.insert arg operand m
  ) Map.empty (zip (AST.parameters f) operands)

  retVal <- foldM (\r i -> case i of
    AST.RetInst _ val _ -> return $ Just val
    _ -> do
      newI <- IRB.cloneInstruction i
      IRB.emitInstr (AST.typeOf newI) newI
      let vMap' = Map.insert i newI vMap
      IRB.remapInstructionInPlace newI vMap' [IRB.RF_NoModuleLevelChanges, IRB.RF_IgnoreMissingLocals]
      case newI of
        AST.CallInst _ _ (Right callee) _ _ _ -> do
          assert (AST.isIntrinsic callee) "Called function must be an intrinsic"
          let m = AST.moduleDefinitions $ IRB.getModule
          intrinsicDecl <- IRB.emitDefn $ AST.GlobalDefinition $
            AST.Function (AST.linkage callee) (AST.visibility callee) (AST.dllStorageClass callee)
                         (AST.callingConvention callee) (AST.returnAttributes callee) (AST.returnType callee)
                         (AST.name callee) (AST.parameters callee) (AST.functionAttributes callee)
                         (AST.section callee) (AST.comdat callee) (AST.alignment callee)
                         (AST.garbageCollectorName callee) (AST.prefix callee) (AST.basicBlocks callee)
                         (AST.personalityFunction callee) (AST.metadata callee)
          IRB.modifyInstruction newI $ \case
            AST.CallInst t s _ args attrs md -> AST.CallInst t s (Right intrinsicDecl) args attrs md
            other -> other
        _ -> return ()
      return r
  ) Nothing (AST.basicBlockInstructions bb)

  case retVal of
    Just val -> case Map.lookup val vMap of
      Just output -> return output
      Nothing -> error "Return value not found in value map"
    Nothing -> error "Wrapper not returning explicitly"

assert :: Bool -> String -> IRB.IRBuilder ()
assert cond msg = unless cond $ error msg

File: rust/src/vectorize/intrinsic_builder.rs
use llvm_sys::prelude::*;
use llvm_sys::core::*;
use std::ffi::CString;
use std::collections::HashMap;

pub struct IntrinsicBuilder {
    inst_wrappers: LLVMModuleRef,
    builder: LLVMBuilderRef,
}

impl IntrinsicBuilder {
    pub unsafe fn create(&self, name: &str, operands: &[LLVMValueRef], imm8: u8) -> LLVMValueRef {
        let wrapper_name = format!("intrinsic_wrapper_{}_{}", name, imm8);
        let wrapper_name_c = CString::new(wrapper_name).unwrap();
        let f = LLVMGetNamedFunction(self.inst_wrappers, wrapper_name_c.as_ptr());
        assert!(!f.is_null(), "Intrinsic wrapper undefined.");

        let bb = LLVMGetFirstBasicBlock(f);
        assert!(!bb.is_null() && LLVMGetNextBasicBlock(bb).is_null(), 
                "Intrinsic Wrapper should have a single basic block");

        let num_args = LLVMCountParams(f);
        assert_eq!(operands.len(), num_args as usize);

        let mut v_map = HashMap::new();
        for (i, &arg) in LLVMGetParams(f).iter().enumerate() {
            assert!(LLVMCastIsValid(LLVMBitCast, operands[i], LLVMTypeOf(arg)) != 0,
                    "Invalid input type");
            let operand = LLVMBuildBitCast(self.builder, operands[i], LLVMTypeOf(arg), b"\0".as_ptr() as *const _);
            v_map.insert(arg, operand);
        }

        let mut ret_val = None;
        let mut inst = LLVMGetFirstInstruction(bb);
        while !inst.is_null() {
            if LLVMGetInstructionOpcode(inst) == LLVMRet {
                ret_val = Some(LLVMGetReturnValue(inst));
                break;
            }
            let new_i = LLVMInstructionClone(inst);
            LLVMInsertIntoBuilder(self.builder, new_i);
            v_map.insert(inst, new_i);
            LLVMRemapInstructionInPlace(new_i, |old_val| {
                v_map.get(&old_val).cloned().unwrap_or(old_val)
            }, LLVMRemapInstructionFlags::LLVMRemapIgnoreMissingLocals as u32);

            if LLVMGetInstructionOpcode(new_i) == LLVMCall {
                let callee = LLVMGetCalledValue(new_i);
                if LLVMIsAFunction(callee) != std::ptr::null_mut() {
                    let callee_func = LLVMValueAsFunction(callee);
                    assert!(LLVMIsFunctionVarArg(callee_func) != 0, "Called function must be an intrinsic");
                    let m = LLVMGetGlobalParent(LLVMGetInstructionParent(new_i));
                    let intrinsic_decl = LLVMAddFunction(m, LLVMGetValueName(callee_func), LLVMGetElementType(LLVMTypeOf(callee_func)));
                    LLVMSetFunctionCallConv(intrinsic_decl, LLVMGetFunctionCallConv(callee_func));
                    LLVMSetInstructionCallConv(new_i, LLVMGetFunctionCallConv(callee_func));
                    LLVMSetCalledValue(new_i, intrinsic_decl);
                }
            }

            inst = LLVMGetNextInstruction(inst);
        }

        let ret_val = ret_val.expect("Wrapper not returning explicitly");
        let output = v_map.get(&ret_val).expect("Return value not found in value map");
        *output
    }
}